.agents/architecture/ADR-016-github-mcp-agent-isolation.md

```diff
@@ -0,0 +1,423 @@
+---
+status: proposed
+date: 2025-12-23
+decision-makers: User, Architect Agent
+consulted: Analyst (Session 80), Independent-Thinker
+informed: Implementer, DevOps, Orchestrator
+---
+
+# ADR-016: GitHub MCP Server with Agent Isolation Pattern
+
+## Context and Problem Statement
+
+Claude Code's interaction with GitHub currently uses PowerShell wrapper scripts that suffer from significant latency overhead. Each `pwsh` invocation from the Bash tool spawns a new process:
+
+- **Without -NoProfile**: 1,044ms per call
+- **With -NoProfile** (PR #285): 183ms per call
+
+For PR review workflows that require 5-10 GitHub API calls, this adds 1-4 seconds of latency per interaction. The question is: what architecture best addresses this performance problem while maintaining code quality, security, and maintainability?
+
+## Decision Drivers
+
+1. **Performance**: Minimize latency for GitHub operations
+2. **Context Efficiency**: Prevent tool explosion in main Claude Code context
+3. **Maintainability**: Reduce custom code maintenance burden
+4. **Security**: Maintain secure credential handling
+5. **Platform Portability**: Support Windows, Linux, macOS
+6. **ADR-005 Compliance**: Respect PowerShell-only scripting standard where appropriate
+7. **Capability Coverage**: Support all current use cases (PRs, issues, reviews, reactions, labels)
+
+## Considered Options
+
+### Option A: PowerShell Wrappers (Current State)
+
+**Description**: 20 PowerShell scripts in `.claude/skills/github/scripts/` called via `pwsh -NoProfile`.
+
+**Performance**: 183-416ms per call (process spawn overhead)
+
+**Capabilities**:
+
+| Operation | Supported | Implementation |
+|-----------|-----------|----------------|
+| PR metadata | Yes | Get-PRContext.ps1 |
+| PR review comments | Yes | Get-PRReviewComments.ps1 |
+| Reply to review | Yes | Post-PRCommentReply.ps1 |
+| Resolve thread | Yes | Resolve-PRReviewThread.ps1 (GraphQL) |
+| Add reaction | Yes | Add-CommentReaction.ps1 |
+| Labels | Yes | Set-IssueLabels.ps1 |
+| Milestones | Yes | Set-IssueMilestone.ps1 |
+| Issue operations | Yes | Get-IssueContext.ps1, etc. |
+| GraphQL mutations | Yes | Direct gh api graphql calls |
+
+**Pros**:
+- Full capability coverage (100% of use cases)
+- Tested and production-proven
+- Follows ADR-005 PowerShell-only standard
+- Type-safe object pipelines
+- Structured error handling
+
+**Cons**:
+- 183-416ms latency per call (unavoidable spawn overhead)
+- No native server mode in PowerShell
+- 20 scripts to maintain
+
+### Option B: gh CLI Rewrite (Issue #286)
```

Comment:
@rjmurillo-bot Issue #286 needs to be closed as not planned. It is superceded by #289